File: joiner-spring/src/main/java/cz/encircled/joiner/spring/SpringJoinerRepository.java
package cz.encircled.joiner.spring;

import cz.encircled.joiner.core.JoinerRepository;
import cz.encircled.joiner.query.JoinerQuery;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Entity repository with spring-data integration
 *
 * @author Kisel on 31.10.2016.
 */
public interface SpringJoinerRepository<T> extends JoinerRepository<T> {

    /**
     * This "find" method support spring {@link Pageable} and converts result to spring {@link Page}
     * <p>
     *     {@link PageableFeature} is used
     * </p>
     *
     * @see PageableFeature
     * @param request query request
     * @param <R> return type
     * @return find result converted to spring page
     */
    <R, U extends T> Page<R> findPage(JoinerQuery<U, R> request, Pageable pageable);

}

File: joiner-spring/src/main/java/cz/encircled/joiner/spring/SpringJoinerRepository.kt
package cz.encircled.joiner.spring

import com.mysema.query.types.Expression
import cz.encircled.joiner.core.Joiner
import cz.encircled.joiner.core.JoinerRepository
import cz.encircled.joiner.query.Q
import org.springframework.beans.factory.annotation.Autowired

/**
 * Parent class for repositories with Joiner support within spring context.

 * @author Vlad on 14-Aug-16.
 */
abstract class SpringJoinerRepository<T> : JoinerRepository<T> {

    @Autowired
    protected var delegate: Joiner? = null

    override fun find(request: Q<T>): List<T> {
        return delegate!!.find(request)
    }

    override fun <P> find(request: Q<T>, projection: Expression<P>): List<P> {
        return delegate!!.find(request, projection)
    }

    override fun findOne(request: Q<T>): T? {
        return delegate!!.findOne(request)
    }

    override fun <P> findOne(request: Q<T>, projection: Expression<P>): P? {
        return delegate!!.findOne(request, projection)
    }

}